File: packages/shell-protocol/src/messages.ts
import { z } from 'zod';

export const SessionInfoSchema = z.object({
  id: z.string(),
  name: z.string(),
  shell: z.string(),
  cwd: z.string(),
  started_at: z.string(),
  // New optional fields for enhanced session tracking
  state: z.enum(['running', 'exited', 'closed', 'error']).optional(),
  foreground_process: z.string().optional(),       // e.g., "claude", "vim", "zsh"
  last_activity_at: z.string().optional(),         // ISO timestamp
  exit_code: z.number().optional(),                // set when state="exited"
  // Docker container fields
  container_id: z.string().optional(),
  container_name: z.string().optional(),
  container_image: z.string().optional(),
  // SSH connection fields
  ssh_connection_id: z.string().optional(),
  ssh_host: z.string().optional(),
  ssh_user: z.string().optional(),
});

export type SessionInfo = z.infer<typeof SessionInfoSchema>;

// Base schema for messages that include a session_id
const WithSessionId = z.object({ session_id: z.string() });

// Mirrors server/src/settings.rs's `TerminalSettings` (serialized camelCase).
// This is the wire shape sent/received over GetSettings/PutSettings/SettingsData
// — distinct from web/src/lib/settingsStore.svelte.ts's client-local
// `TerminalSettings` type, which already covers a different (UI-only) set of
// fields; reconciling the two is a pre-existing drift issue, out of scope here.
const TerminalSettingsSchema = z.object({
  fontSize: z.number(),
  fontFamily: z.string(),
  fontColor: z.string(),
  backgroundColor: z.string(),
  cursorStyle: z.string(),
  cursorBlink: z.boolean(),
  lineHeight: z.number(),
  defaultCwd: z.string(),
});

// Mirrors server/src/workspace.rs's `SplitNode`/`Tab`/`Workspace`/
// `LayoutTemplate`/`WorkspaceState` (serialized camelCase; SplitNode is a
// `type`-tagged union, "pane" | "split"). Used by GetWorkspaceState/
// PutWorkspaceState/WorkspaceStateData — distinct from the existing
// LoadWorkspace/SaveWorkspace/WorkspaceData messages, which persist opaque
// per-client JSON to a different file and are unrelated to this shape.
const SplitNodeSchema: z.ZodType<unknown> = z.lazy(() =>
  z.union([
    z.object({ type: z.literal('pane'), id: z.string(), sessionId: z.string().nullable() }),
    z.object({
      type: z.literal('split'),
      id: z.string(),
      direction: z.enum(['horizontal', 'vertical']),
      children: z.array(SplitNodeSchema),
      ratios: z.array(z.number()),
    }),
  ])
);

const TabSchema = z.object({ id: z.string(), name: z.string(), root: SplitNodeSchema });

const WorkspaceSchema = z.object({ tabs: z.array(TabSchema), activeTabId: z.string() });

const LayoutTemplateSchema = z.object({
  id: z.string(),
  name: z.string(),
  root: SplitNodeSchema,
  createdAt: z.string(),
});

const WorkspaceStateSchema = z.object({
  workspace: WorkspaceSchema,
  templates: z.array(LayoutTemplateSchema),
});

export const ClientMessageSchema = z.union([
  z.object({ type: z.literal('auth'), token: z.string(), protocol_version: z.string().optional() }),
  z.object({ type: z.literal('list_sessions') }),
  z.object({
    type: z.literal('create_session'),
    cwd: z.string(),
    shell: z.string(),
    env: z.record(z.string(), z.string()),
    cols: z.number(),
    rows: z.number(),
    container_id: z.string().optional(),
    ssh_connection_id: z.string().optional(),
  }),
  WithSessionId.extend({ type: z.literal('attach'), mode: z.enum(['mirror', 'exclusive']) }),
  WithSessionId.extend({ type: z.literal('input'), data: z.string() }),
  WithSessionId.extend({ type: z.literal('resize'), cols: z.number(), rows: z.number() }),
  WithSessionId.extend({ type: z.literal('rename_session'), new_name: z.string() }),
  WithSessionId.extend({ type: z.literal('kill_session') }),
  // Reply to a server-side `EditRequest`. `id` echoes the correlation
  // token from the request. `cancelled=true` signals the user dismissed
  // the modal; `contents` is ignored in that case. The server frames
  // this and writes it to the session PTY for the wrapper script to read.
  WithSessionId.extend({
    type: z.literal('edit_reply'),
    id: z.string(),
    contents: z.string(),
    cancelled: z.boolean(),
  }),
  z.object({ type: z.literal('auth_password'), username: z.string(), password: z.string() }),
  z.object({ type: z.literal('auth_pubkey_init'), username: z.string(), pubkey: z.string() }),
  z.object({ type: z.literal('auth_pubkey_verify'), signature: z.string(), algorithm: z.string() }),
  z.object({ type: z.literal('auth_token'), token: z.string() }),
  z.object({ type: z.literal('refresh_token'), refresh_token: z.string() }),
  z.object({ type: z.literal('list_containers'), ssh_connection_id: z.string().optional() }),
  z.object({ type: z.literal('list_ssh_connections') }),
  z.object({
    type: z.literal('add_ssh_connection'),
    name: z.string(),
    host: z.string(),
    user: z.string(),
    port: z.number(),
  }),
  z.object({
    type: z.literal('update_ssh_connection'),
    id: z.string(),
    name: z.string(),
    host: z.string(),
    user: z.string(),
    port: z.number(),
  }),
  z.object({ type: z.literal('remove_ssh_connection'), id: z.string() }),
  z.object({ type: z.literal('import_ssh_config') }),
  z.object({ type: z.literal('save_workspace'), workspace: z.record(z.string(), z.unknown()) }),
  z.object({ type: z.literal('load_workspace') }),
  z.object({ type: z.literal('get_workspace_state') }),
  z.object({ type: z.literal('put_workspace_state'), state: WorkspaceStateSchema }),
  z.object({ type: z.literal('get_settings') }),
  z.object({ type: z.literal('put_settings'), settings: TerminalSettingsSchema }),
  z.object({ type: z.literal('get_themes') }),
  z.object({ type: z.literal('put_themes'), value: z.unknown() }),
  z.object({ type: z.literal('get_tags') }),
  z.object({ type: z.literal('put_tags'), value: z.unknown() }),
]);

export type ClientMessage = z.infer<typeof ClientMessageSchema>;

export const ServerMessageSchema = z.union([
  z.object({ type: z.literal('AuthOk'), token: z.string(), expires: z.string(), protocol_version: z.string().optional(), refresh_token: z.string().optional() }),
  z.object({ type: z.literal('AuthChallenge'), nonce: z.string() }),
  z.object({ type: z.literal('SessionList'), sessions: z.array(SessionInfoSchema) }),
  WithSessionId.extend({ type: z.literal('Output'), data: z.string() }),
  WithSessionId.extend({ type: z.literal('SessionClosed') }),
  z.object({ type: z.literal('Error'), message: z.string(), error_code: z.string().optional() }),
  z.object({ type: z.literal('PairResponse'), code: z.string(), expiry_secs: z.number() }),
  z.object({ type: z.literal('Shutdown'), reason: z.string() }),
  WithSessionId.extend({ type: z.literal('ForegroundChanged'), process_name: z.string().nullable() }),
  WithSessionId.extend({ type: z.literal('SessionActivity'), activity_type: z.enum(['activity', 'bell', 'silence']) }),
  WithSessionId.extend({ type: z.literal('SessionExited'), exit_code: z.number().nullable() }),
  WithSessionId.extend({ type: z.literal('CwdChanged'), cwd: z.string() }),
  // Tool-action events: a program in the remote session emitted an OSC
  // sequence asking the local client to perform a native action on the
  // user's computer. `data` and `url` are already decoded on the server
  // (the OSC wire encoding uses base64; we don't re-encode for JSON).
  WithSessionId.extend({ type: z.literal('ClipboardWrite'), data: z.string() }),
  WithSessionId.extend({ type: z.literal('OpenUrl'), url: z.string() }),
  // A program in the remote session emitted an OSC 7777 `edit_request`,
  // asking the user to edit a file locally and send the result back.
  // `id` is an opaque correlation token the wrapper script chose; the
  // client must echo it back in the matching `edit_reply`. `filename`
  // is the original path on the remote (display only — the local edit
  // round-trips through the PTY, not the filesystem). `contents` is the
  // current file contents the modal should pre-fill.
  WithSessionId.extend({
    type: z.literal('EditRequest'),
    id: z.string(),
    filename: z.string(),
    contents: z.string(),
  }),
  z.object({ type: z.literal('WorkspaceData'), workspace: z.record(z.string(), z.unknown()).nullable() }),
  z.object({ type: z.literal('SettingsData'), settings: TerminalSettingsSchema }),
  z.object({ type: z.literal('ThemesData'), value: z.unknown().nullable() }),
  z.object({ type: z.literal('TagsData'), value: z.unknown().nullable() }),
  z.object({ type: z.literal('WorkspaceStateData'), state: WorkspaceStateSchema }),
  z.object({
    type: z.literal('ContainerList'),
    containers: z.array(z.object({
      id: z.string(),
      name: z.string(),
      image: z.string(),
      status: z.string(),
      state: z.string(),
    })),
  }),
  z.object({
    type: z.literal('SshConnectionList'),
    connections: z.array(z.object({
      id: z.string(),
      name: z.string(),
      host: z.string(),
      user: z.string(),
      port: z.number(),
    })),
  }),
  z.object({
    type: z.literal('SshConfigImportResult'),
    hosts: z.array(z.object({
      name: z.string(),
      hostname: z.string().optional(),
      user: z.string().optional(),
      port: z.number().optional(),
    })),
  }),
]);

export type ServerMessage = z.infer<typeof ServerMessageSchema>;

export interface ContainerInfo {
  id: string;
  name: string;
  image: string;
  status: string;
  state: string;
}

export interface SshConnectionInfo {
  id: string;
  name: string;
  host: string;
  user: string;
  port: number;
}

export interface SshConfigHost {
  name: string;
  hostname?: string;
  user?: string;
  port?: number;
}

File: server/src/messages.rs
//! Protocol message types for client-server communication.
//!
//! This module re-exports core session messages from `terminar_core`
//! and defines workspace messages that remain server-specific.
//!
//! The server uses wrapper enums `ClientMessage` and `ServerMessage` that
//! combine core + server-specific variants for unified dispatch.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// Re-export core types
pub use terminar_core::messages::SessionInfo;

/// Messages sent from clients (VS Code extension, web frontend) to the server.
///
/// Serialized with a `"type"` field in `snake_case` (e.g., `"create_session"`).
#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type")]
#[serde(rename_all = "snake_case")]
pub enum ClientMessage {
    // === Core session messages (mirrored from terminar_core) ===
    /// Authenticate a WebSocket connection. Must be the first message on remote connections.
    Auth {
        /// The API token (UUID).
        token: String,
        /// Protocol version the client supports (e.g., "0.2.0").
        #[serde(skip_serializing_if = "Option::is_none")]
        protocol_version: Option<String>,
    },
    /// Request the list of all active sessions.
    ListSessions,
    /// Create a new terminal session with a PTY process.
    CreateSession {
        /// Working directory for the shell. Empty or `"/"` defaults to `$HOME`.
        cwd: String,
        /// Shell binary path. Must be in the allowed whitelist. Empty defaults to `$SHELL`.
        shell: String,
        /// Environment variables to set in the shell. Dangerous variables are filtered.
        env: HashMap<String, String>,
        /// Terminal width in columns (clamped to 1-500).
        cols: u16,
        /// Terminal height in rows (clamped to 1-500).
        rows: u16,
        /// Docker container ID for container sessions. When set, the session runs
        /// `docker exec` inside the specified container instead of a local shell.
        #[serde(skip_serializing_if = "Option::is_none")]
        container_id: Option<String>,
        /// SSH connection ID for SSH sessions. When set, the session runs
        /// `ssh user@host` to the remote host instead of a local shell.
        #[serde(skip_serializing_if = "Option::is_none")]
        ssh_connection_id: Option<String>,
    },
    /// Attach to a session to receive its output stream.
    Attach {
        /// UUID of the session to attach to.
        session_id: String,
        /// Attach mode (currently only `"mirror"` is supported).
        mode: String,
    },
    /// Send keyboard input to a session's PTY.
    Input {
        /// UUID of the target session.
        session_id: String,
        /// Raw input data including control characters.
        data: String,
    },
    /// Resize a session's terminal dimensions.
    Resize {
        /// UUID of the target session.
        session_id: String,
        /// New terminal width in columns (clamped to 1-500).
        cols: u16,
        /// New terminal height in rows (clamped to 1-500).
        rows: u16,
    },
    /// Rename an existing session.
    RenameSession {
        /// UUID of the session to rename.
        session_id: String,
        /// New display name.
        new_name: String,
    },
    /// Terminate a session and its PTY process.
    KillSession {
        /// UUID of the session to kill.
        session_id: String,
    },
    /// Client-supplied reply for an `EditRequest` that was sent to the user.
    /// When `cancelled == false`, `contents` is the text the user typed into
    /// the modal editor and the server will write a framed response to the
    /// session's PTY stdin so the remote wrapper script can decode and save
    /// it. When `cancelled == true`, `contents` is ignored and the server
    /// signals cancellation to the wrapper.
    EditReply {
        session_id: String,
        id: String,
        contents: String,
        cancelled: bool,
    },

    // === Docker container messages ===
    /// Request the list of running Docker containers. When `ssh_connection_id`
    /// is set, queries the Docker daemon on the remote host via
    /// `DOCKER_HOST=ssh://...` instead of the local daemon.
    ListContainers {
        #[serde(skip_serializing_if = "Option::is_none")]
        ssh_connection_id: Option<String>,
    },

    // === SSH connection messages ===
    /// Request the list of saved SSH connections.
    ListSshConnections,
    /// Add a new SSH connection.
    AddSshConnection {
        name: String,
        host: String,
        user: String,
        port: u16,
    },
    /// Update an existing SSH connection.
    UpdateSshConnection {
        id: String,
        name: String,
        host: String,
        user: String,
        port: u16,
    },
    /// Remove an SSH connection.
    RemoveSshConnection { id: String },
    /// Parse ~/.ssh/config and return importable hosts.
    ImportSshConfig,

    // === Server-specific messages (workspace) ===
    /// Save workspace data (session layout, splits, tabs, etc.).
    SaveWorkspace { workspace: serde_json::Value },
    /// Load previously saved workspace data.
    LoadWorkspace,
}

/// Messages sent from the server to connected clients.
///
/// Serialized with a `"type"` field in `PascalCase` (e.g., `"SessionList"`).
#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum ServerMessage {
    // === Core session messages (mirrored from terminar_core) ===
    /// List of all active sessions. Sent in response to list/create/rename/kill operations.
    SessionList { sessions: Vec<SessionInfo> },
    /// Terminal output data from a session.
    Output { session_id: String, data: String },
    /// A session has been closed (killed or shell exited).
    SessionClosed { session_id: String },
    /// An error occurred processing a client message.
    Error {
        message: String,
        /// Machine-readable error code (e.g., "SESSION_NOT_FOUND", "AUTH_FAILED").
        #[serde(skip_serializing_if = "Option::is_none")]
        error_code: Option<String>,
    },
    /// Notification that the foreground process in a session has changed.
    ForegroundChanged {
        session_id: String,
        process_name: Option<String>,
    },
    /// Notification of session activity (output, bell, or silence marker).
    SessionActivity {
        session_id: String,
        activity_type: String, // "activity", "bell", "silence"
    },
    /// Notification that a session's shell has exited.
    SessionExited {
        session_id: String,
        exit_code: Option<i32>,
    },
    /// Notification that the current working directory in a session has changed.
    CwdChanged { session_id: String, cwd: String },
    /// Notification that a program in a session emitted an OSC 52 clipboard
    /// write. The client should copy `data` to the user's local clipboard.
    /// `data` is the text to copy as UTF-8 (the base64 wire encoding used by
    /// OSC 52 has already been decoded on the server).
    ClipboardWrite { session_id: String, data: String },
    /// Notification that a program in a session requested a URL be opened
    /// in the user's local default browser (via OSC 7777 `open_url`).
    OpenUrl { session_id: String, url: String },
    /// Notification that a program in a session requested the user edit a
    /// file locally and send the result back (via OSC 7777 `edit_request`).
    EditRequest {
        session_id: String,
        id: String,
        filename: String,
        contents: String,
    },

    // === Docker container messages ===
    /// List of running Docker containers.
    ContainerList { containers: Vec<ContainerInfo> },

    // === SSH connection messages ===
    /// List of saved SSH connections.
    SshConnectionList { connections: Vec<SshConnectionInfo> },
    /// Result of parsing ~/.ssh/config.
    SshConfigImportResult { hosts: Vec<SshConfigHost> },

    // === Server-specific messages (auth/workspace) ===
    /// Authentication succeeded. Contains the token and protocol version.
    AuthOk {
        token: String,
        expires: String,
        /// Protocol version the server supports (e.g., "0.2.0").
        #[serde(skip_serializing_if = "Option::is_none")]
        protocol_version: Option<String>,
    },
    /// Server is shutting down gracefully. Clients should reconnect later.
    Shutdown { reason: String },
    /// Response to LoadWorkspace with saved workspace data.
    WorkspaceData {
        workspace: Option<serde_json::Value>,
    },
}

/// Information about a running Docker container.
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ContainerInfo {
    /// Container ID (short form).
    pub id: String,
    /// Container name.
    pub name: String,
    /// Container image.
    pub image: String,
    /// Human-readable status (e.g., "Up 2 hours").
    pub status: String,
    /// Container state (e.g., "running").
    pub state: String,
}

/// A saved SSH connection.
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SshConnectionInfo {
    pub id: String,
    pub name: String,
    pub host: String,
    pub user: String,
    pub port: u16,
}

/// A host entry parsed from ~/.ssh/config.
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SshConfigHost {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hostname: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_client_message_auth() {
        let msg = ClientMessage::Auth {
            token: "abc".into(),
            protocol_version: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert_eq!(json, r#"{"type":"auth","token":"abc"}"#);

        let deserialized: ClientMessage = serde_json::from_str(&json).unwrap();
        match deserialized {
            ClientMessage::Auth { token, .. } => assert_eq!(token, "abc"),
            _ => panic!("Wrong type"),
        }
    }

    #[test]
    fn test_client_message_list_sessions() {
        let msg = ClientMessage::ListSessions;
        let json = serde_json::to_string(&msg).unwrap();
        assert_eq!(json, r#"{"type":"list_sessions"}"#);
    }

    #[test]
    fn test_client_message_create_session() {
        let mut env = HashMap::new();
        env.insert("FOO".into(), "BAR".into());
        let msg = ClientMessage::CreateSession {
            cwd: "/tmp".into(),
            shell: "sh".into(),
            env,
            cols: 100,
            rows: 50,
            container_id: None,
            ssh_connection_id: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"create_session""#));
        assert!(json.contains(r#""cwd":"/tmp""#));
    }

    #[test]
    fn test_client_message_input() {
        let msg = ClientMessage::Input {
            session_id: "id".into(),
            data: "ls\n".into(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert_eq!(json, r#"{"type":"input","session_id":"id","data":"ls\n"}"#);
    }

    #[test]
    fn test_server_message_session_list() {
        let info = SessionInfo {
            id: "1".into(),
            name: "n".into(),
            shell: "s".into(),
            cwd: "/home".into(),
            started_at: "t".into(),
            state: None,
            foreground_process: None,
            last_activity_at: None,
            exit_code: None,
            container_id: None,
            container_name: None,
            container_image: None,
            ssh_connection_id: None,
            ssh_host: None,
            ssh_user: None,
        };
        let msg = ServerMessage::SessionList {
            sessions: vec![info],
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("SessionList"));
    }

    #[test]
    fn test_server_message_output() {
        let msg = ServerMessage::Output {
            session_id: "1".into(),
            data: "d".into(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert_eq!(json, r#"{"type":"Output","session_id":"1","data":"d"}"#);
    }

    #[test]
    fn test_server_message_error() {
        let msg = ServerMessage::Error {
            message: "err".into(),
            error_code: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert_eq!(json, r#"{"type":"Error","message":"err"}"#);
    }

    #[test]
    fn test_server_message_auth_ok() {
        let msg = ServerMessage::AuthOk {
            token: "tok".into(),
            expires: "900s".into(),
            protocol_version: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"AuthOk""#));
    }

    #[test]
    fn test_server_message_foreground_changed() {
        let msg = ServerMessage::ForegroundChanged {
            session_id: "id1".into(),
            process_name: Some("vim".into()),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"ForegroundChanged""#));
        assert!(json.contains(r#""process_name":"vim""#));
    }

    #[test]
    fn test_server_message_session_exited() {
        let msg = ServerMessage::SessionExited {
            session_id: "id1".into(),
            exit_code: Some(0),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""exit_code":0"#));
    }

    #[test]
    fn test_server_message_cwd_changed() {
        let msg = ServerMessage::CwdChanged {
            session_id: "id1".into(),
            cwd: "/tmp".into(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"CwdChanged""#));
    }

    #[test]
    fn test_server_message_workspace_data() {
        let msg = ServerMessage::WorkspaceData { workspace: None };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"WorkspaceData""#));
    }

    #[test]
    fn test_server_message_edit_request() {
        let msg = ServerMessage::EditRequest {
            session_id: "sess-1".into(),
            id: "req-42".into(),
            filename: "note.txt".into(),
            contents: "hello\nworld".into(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"EditRequest""#));
        assert!(json.contains(r#""session_id":"sess-1""#));
        assert!(json.contains(r#""id":"req-42""#));
        assert!(json.contains(r#""filename":"note.txt""#));
        assert!(json.contains(r#""contents":"hello\nworld""#));

        let deserialized: ServerMessage = serde_json::from_str(&json).unwrap();
        match deserialized {
            ServerMessage::EditRequest {
                session_id,
                id,
                filename,
                contents,
            } => {
                assert_eq!(session_id, "sess-1");
                assert_eq!(id, "req-42");
                assert_eq!(filename, "note.txt");
                assert_eq!(contents, "hello\nworld");
            }
            _ => panic!("Wrong type"),
        }
    }

    #[test]
    fn test_client_message_edit_reply() {
        let msg = ClientMessage::EditReply {
            session_id: "sess-1".into(),
            id: "req-42".into(),
            contents: "edited text".into(),
            cancelled: false,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"edit_reply""#));
        assert!(json.contains(r#""session_id":"sess-1""#));
        assert!(json.contains(r#""id":"req-42""#));
        assert!(json.contains(r#""contents":"edited text""#));
        assert!(json.contains(r#""cancelled":false"#));

        let deserialized: ClientMessage = serde_json::from_str(&json).unwrap();
        match deserialized {
            ClientMessage::EditReply {
                session_id,
                id,
                contents,
                cancelled,
            } => {
                assert_eq!(session_id, "sess-1");
                assert_eq!(id, "req-42");
                assert_eq!(contents, "edited text");
                assert!(!cancelled);
            }
            _ => panic!("Wrong type"),
        }
    }

    #[test]
    fn test_client_message_edit_reply_cancelled() {
        let msg = ClientMessage::EditReply {
            session_id: "sess-1".into(),
            id: "req-42".into(),
            contents: String::new(),
            cancelled: true,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"edit_reply""#));
        assert!(json.contains(r#""cancelled":true"#));
        assert!(json.contains(r#""contents":"""#));

        let deserialized: ClientMessage = serde_json::from_str(&json).unwrap();
        match deserialized {
            ClientMessage::EditReply {
                session_id,
                id,
                contents,
                cancelled,
            } => {
                assert_eq!(session_id, "sess-1");
                assert_eq!(id, "req-42");
                assert_eq!(contents, "");
                assert!(cancelled);
            }
            _ => panic!("Wrong type"),
        }
    }

    #[test]
    fn test_session_info_backward_compat() {
        let old_json = r#"{"id":"1","name":"n","shell":"s","cwd":"/home","started_at":"t"}"#;
        let deserialized: SessionInfo = serde_json::from_str(old_json).unwrap();
        assert_eq!(deserialized.id, "1");
        assert_eq!(deserialized.state, None);
    }
}
